HW3/P5/label_regions.cl

```diff
@@ -80,20 +80,82 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
+
+    /*// part II:
+    // update each label by its parents (only for non-halo vals)
+    if(idx_1D > halo - 1 && idx_1D < buf_w - halo) {
+        for (int row = halo; row < buf_h - halo; row++) {
+            int offset = row * buf_w + idx_1D;
+
+            // only foreground pixels
+            if(buffer[offset] < w * h)
+                buffer[offset] = labels[buffer[offset]];  
+        }
+    }*/
+
+    // part IV: --> single thread version!
+    // Assumption: we do not update halo values
+    int previous_val = -1;
+    int previous_idx = -1;
+    for(int row = halo; row < buf_h - halo; row++)
+        for(int col = halo; col < buf_w - halo; col++) {
+            int offset = row * buf_w + col;
+
+            // only foreground pixels
+            if(buffer[offset] < w * h) 
+                // fetch only global val if index is different from last fetch
+                if(previous_idx != buffer[offset]) {
+                    previous_idx = buffer[offset];
+                    buffer[offset] = labels[buffer[offset]];
+                    previous_val = buffer[offset];
+                }
+                else {
+                    buffer[offset] = previous_val;
+                }
+        }
+
+    barrier(CLK_LOCAL_MEM_FENCE);
     
     // stay in bounds
     if ((x < w) && (y < h)) {
         // CODE FOR PART 1 HERE
         // We set new_label to the value of old_label, but you will need
         // to adjust this for correctness.
+
         new_label = old_label;
+        // compute min only for foreground pixel
+        if(old_label < w * h) {
+            int vert = min(buffer[(buf_y + 1) * buf_w + buf_x], buffer[(buf_y - 1) * buf_w + buf_x]);
+            int horz = min(buffer[buf_y * buf_w + buf_x + 1], buffer[buf_y * buf_w + buf_x - 1]);
+            new_label = min(old_label, min(vert, horz));
```

Comment:
After parts 2 and 4, you should use buffer[buf_w * buf_y + buf_x] instead of old_label.